### ISSUE TITLE
how to define wsgi_module for multiple vhosts 

### ISSUE THREAD

**Issue by hzuluaga** [GrahamDumpleton/mod_wsgi#651]
Hello @GrahamDumpleton 

I am trying to deploy 2 flasks sites (contapermenv and patrimonioenv) on an Apache-WindowsServer as vhosts. Each site is a virtualenv with its own python libraries. I have installed on each virtualenv  "mod_wsgi" using the command: 
````
pip install https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip
````
 
 For instance, for the "contapermenv" site,  the generated lines are:

````
LoadFile "c:/program files/python38/python38.dll"
LoadModule wsgi_module "c:/apache24/htdocs/contapermenv/lib/site-packages/mod_wsgi/server/mod_wsgi.cp38-win_amd64.pyd"
WSGIPythonHome "c:/apache24/htdocs/contapermenv"
````
and for the "patrimonioenv" site, the generated lines are:

````
LoadFile "c:/program files/python38/python38.dll"
LoadModule wsgi_module "c:/apache24/htdocs/patrimonioenv/lib/site-packages/mod_wsgi/server/mod_wsgi.cp38-win_amd64.pyd"
WSGIPythonHome "c:/apache24/htdocs/patrimonioenv"
````

According to what I have found, the returned lines  by running "mod_wsgi-express module-config" have to be included in the httpd.conf file and not inside the httpd-vhosts.conf file.
If I include both set of lines for each site within my httpd.conf and start apache, I get the following message:

````
[Fri Feb 05 12:56:24.012675 2021] [so:warn] [pid 11608:tid 124] AH01574: module wsgi_module is already loaded, skipping
````

My question is, how do I configure apache in order that each vhost uses its own python virtualenv (own libraries and versions) without interfering with the other installed vhosts. 

This is my httpd-vhosts.conf file:

```
Listen 80
Listen 8080

<VirtualHost *:80>
        
        ServerName 192.168.218.129:80
        WSGIScriptAlias / "C:/Apache24/htdocs/contapermenv/wsgi.py"
        DocumentRoot "C:/Apache24/htdocs/contapermenv"
        <Directory "C:/Apache24/htdocs/contapermenv">
            <Files wsgi.py>
                 Require all granted
            </Files>
        </Directory>

        Alias /static "C:/Apache24/htdocs/contapermenv/defaultapp/static/"
        <Directory "C:/Apache24/htdocs/contapermenv/defaultapp/static/">
              Require all granted
        </Directory>
        ErrorLog "C:/Apache24/htdocs/contapermenv/logs/error.log"
        CustomLog "C:/Apache24/htdocs/contapermenv/logs/access.log" common
</VirtualHost>


<VirtualHost *:8080>
        
        ServerName  192.168.218.129:8080
        WSGIScriptAlias / "C:/Apache24/htdocs/patrimonioenv/wsgi.py"
        DocumentRoot "C:/Apache24/htdocs/patrimonioenv"
        <Directory "C:/Apache24/htdocs/patrimonioenv">
            <Files wsgi.py>
                 Require all granted
            </Files>
        </Directory>

        Alias /static "C:/Apache24/htdocs/patrimonioenv/defaultapp/static/"
        <Directory "C:/Apache24/htdocs/patrimonioenv/defaultapp/static/">
              Require all granted
        </Directory>
        ErrorLog "C:/Apache24/htdocs/patrimonioenv/logs/error.log"
        CustomLog "C:/Apache24/htdocs/patrimonioenv/logs/access.log" common
</VirtualHost>
```

If I start apache with this configuration , and  visit first , let's say "localhost:8080" , the "patrimonio" site is shown , but if I the go to  "localhost" , the same site is displayed , and not the "contaperm", which is the one defined on  port 80.

If , I restart apache , and then visit first "localhost", the "contaperm"site is shown , but if I go to  localhost:8080 , it keeps showing "contaperm" and not "patrimonio", which is the one defined on that port.

I appreciate any help or hint you could give me.

Thx

**Comment by GrahamDumpleton** [comment]
You can't have duplicates of ``LoadModule`` for ``wsgi_module`` as the first will win.

You can't have duplicates of ``WGSIPythonHome`` as the last will win.

Getting separation on Windows (since don't have daemon mode) is tricky and not always possible depending on what third party Python packages you are using.

Steps you have to use are:

* Create a separate virtual environment to install only mod_wsgi.

* Add ``mod_wsgi-express module-config`` output from this separate virtual environment into Apache configuration (only this one).

* In ``wsgi.py`` file for each application you need to use the ``activate_this.py`` method to activate their respective Python virtual environments. You can't declare the location of Python virtual environment in the Apache configuration in this specific situation. See mention of ``activate_this.py`` in https://modwsgi.readthedocs.io/en/master/user-guides/virtual-environments.html#virtual-environments

* In the Apache configuration then use ``application-group`` option to ``WSGIScriptAlias`` to force each WSGI application into a separate Python interpreter context. The deferred activation of the virtual environment from the ``wsgi.py`` file means they should each have separate virtual environment since being told to use different application group (Python sub interpreter context).

```
WSGIScriptAlias / "C:/Apache24/htdocs/contapermenv/wsgi.py" application-group=contapermenv
```

**Comment by hzuluaga** [comment]
@GrahamDumpleton  , thank you for you quick and complete response.
I followed your recommendations and I  created an empty virtualenv app in order to use it as the default   mod_wsgi  source. Its name is "modwsqiappenv"

I added the output of "mod_wsgi-express module-config" command to the httpd.conf file:

````
Listen 80
Listen 8080

LoadFile "c:/program files/python38/python38.dll"
LoadModule wsgi_module "c:/apache24/htdocs/modwsqiappenv/lib/site-packages/mod_wsgi/server/mod_wsgi.cp38-win_amd64.pyd"
WSGIPythonHome "c:/apache24/htdocs/modwsqiappenv"
````

I also added the WSGIScriptAlias directive with the application-group option to each virtual host on the httpd-vhosts.conf file:

````

<VirtualHost *:80>
        
        ServerName 192.168.218.129:80
        WSGIScriptAlias / "C:/Apache24/htdocs/contapermenv/wsgi.py"  application-group=contapermenv
        DocumentRoot "C:/Apache24/htdocs/contapermenv"
        <Directory "C:/Apache24/htdocs/contapermenv">
            <Files wsgi.py>
                 Require all granted
            </Files>
        </Directory>

        Alias /static "C:/Apache24/htdocs/contapermenv/defaultapp/static/"
        <Directory "C:/Apache24/htdocs/contapermenv/defaultapp/static/">
              Require all granted
        </Directory>
        ErrorLog "C:/Apache24/htdocs/contapermenv/logs/error.log"
        CustomLog "C:/Apache24/htdocs/contapermenv/logs/access.log" common
</VirtualHost>


<VirtualHost *:8080>
        
        ServerName  192.168.218.129:8080
        WSGIScriptAlias / "C:/Apache24/htdocs/patrimonioenv/wsgi.py"  application-group=patrimonioenv
        DocumentRoot "C:/Apache24/htdocs/patrimonioenv"
        <Directory "C:/Apache24/htdocs/patrimonioenv">
            <Files wsgi.py>
                 Require all granted
            </Files>
        </Directory>

        Alias /static "C:/Apache24/htdocs/patrimonioenv/defaultapp/static/"
        <Directory "C:/Apache24/htdocs/patrimonioenv/defaultapp/static/">
              Require all granted
        </Directory>
        ErrorLog "C:/Apache24/htdocs/patrimonioenv/logs/error.log"
        CustomLog "C:/Apache24/htdocs/patrimonioenv/logs/access.log" common
</VirtualHost>
````

I also added the call to the  "activate_this.py" method  in each wsgi.py file for each application , as follows:

this is sample of the wsgi.py file for "contapermenv" site:(the one for the other site-app "patrimonioenv" is similar). I had to add the site_packages because I was getting and error when starting apache that said  "No flask module is found", it seemed no python packages or modules were detected so I forced it by adding the packages directory. 
I also had to change  the path to the activate_this.py file , because on windows it is located inside "Scripts" and not "bin". 
````
python_home = 'C:/Apache24/htdocs/contapermenv'
import sys
import site

# Calculate path to site-packages directory.

python_version = '.'.join(map(str, sys.version_info[:2]))
site_packages = python_home + '/Lib/site-packages' 

# Add the site-packages directory.

site.addsitedir(site_packages)

activate_this = python_home + '/Scripts/activate_this.py'
with open(activate_this) as infile:
    exec(infile.read())


sys.path.insert(0,"C:/Apache24/htdocs/contapermenv")
from init import app as application

````
 
I then start apache ...and unfortunately  it never starts, it got stuck, the prompt never returns and there are no errors in the apache log file.

<img width="1306" alt="Screen Shot 2021-02-09 at 12 08 52" src="https://user-images.githubusercontent.com/5191016/107408601-2e2bc500-6ad9-11eb-9ea0-bffd22e3a1e5.png">

```
Starting the 'Apache2.4' service
[Tue Feb 09 12:06:33.623395 2021] [core:warn] [pid 12016:tid 600] AH00098: pid file C:/Apache24/logs/httpd.pid overwritten -- Unclean shutdown of previous Apache run?
[Tue Feb 09 12:06:33.623395 2021] [mpm_winnt:notice] [pid 12016:tid 600] AH00455: Apache/2.4.46 (Win64) mod_wsgi/4.8.0 Python/3.8 configured -- resuming normal operations
[Tue Feb 09 12:06:33.623395 2021] [mpm_winnt:notice] [pid 12016:tid 600] AH00456: Apache Lounge VS16 Server built: Oct  2 2020 11:45:39
[Tue Feb 09 12:06:33.623395 2021] [core:notice] [pid 12016:tid 600] AH00094: Command line: 'C:\\Apache24\\bin\\httpd.exe -d C:/Apache24'
[Tue Feb 09 12:06:33.623395 2021] [mpm_winnt:notice] [pid 12016:tid 600] AH00418: Parent: Created child process 7672

```

What am I missing?

My client wants me to deploy both sites-apps on this windows server-apache  environment , but it has been really hard . Do you think I should go for Linux-apache or Linux-Nginx instead ? 

Thank you very much again for all your help.

regards

**Comment by GrahamDumpleton** [comment]
What third party Python packages are you using? Are you using ``numpy`` or any other packages often used with data science tasks? The ``numpy`` package and others dependent on it will not work in Python sub interpreters and often will lock up Python when used. As such they will not work in this configuration which is dependent on sub interpreters.

**Comment by hzuluaga** [comment]
Hello @GrahamDumpleton 

Indeed , in both apps I use libraries such as pandas , which in turn uses numpy. I uninstalled theses packages to see whether the problem will be gone, and  indeed, apache is no longer blocked and starts ok.

I will go for another option such as linux-nginx , which I have already used and it works perfectly , all virtual hosts with their own  virtualenv treated independently, but I have to convince our client first, to change its servers infrastructure..not easy task !

Thank you very  @GrahamDumpleton  for being so reactive at answering and solving any question or problem we encounter and sorry to bother you with questions that u might have already answered hundreds of times before.

Best regards
